src/r1cs.rs

```diff
@@ -354,14 +385,22 @@ impl<G: Group> R1CSShape<G> {
     // check if the number of variables are as expected, then
     // we simply set the number of constraints to the next power of two
     if self.num_vars == m {
-      return R1CSShape {
+      let r1cs_shape = R1CSShape {
         num_cons: m,
         num_vars: m,
         num_io: self.num_io,
         A: self.A.clone(),
         B: self.B.clone(),
         C: self.C.clone(),
+        // TODO: at this point within the proving pipeline, we assume that `R1CSShape::pad`
+        // is called only in the ppsnark/snark setup process. Thus, the digest's use as a check
+        // to prevent drifting public params is not a concern, and we can avoid doing a massive
+        // recomputation for a new digest that will never be used. However, this is ad-hoc and
+        // we should find a nicer solution
+        digest: self.digest,
```

Comment:
Another concern. As I wrote this, I realized that currently `master` doesn't take into this concern either, but it is more subtle. After padding, the digest of the public params will not match the new padded shape. Looking at the setup code in `snark.rs`/`ppsnark.rs`, I see that we commit again to the padded shape though, so this shouldn't be an issue, right? 

cc: @huitseeker

Reply:
The point of the builder pattern I produced in #31
Was to make sure you'd run all the initialization steps before getting the object, and therefore did not have to check an object to see if it had been padded or not. Did we forget padding as part of this ?

Reply:
I don't understand what you mean. My point is that after the digest is built, we mutate the shape by padding it, and we don't rebuild the new padded shape to update the digest. However, because the new digest wouldn't be used anyway, since the transcript implicitly consumes the digest of the unpadded shape through pp.digest. Hence we can elide the expensive recompute.

However, I don't know if this may raise ways to attack the snark protocol.